kong/clustering/services/sync/hooks.lua

```diff
@@ -0,0 +1,178 @@
+local EMPTY = require("kong.tools.table").EMPTY
+
+
+local ipairs = ipairs
+local ngx_null = ngx.null
+local ngx_log = ngx.log
+local ngx_ERR = ngx.ERR
+local ngx_DEBUG = ngx.DEBUG
+
+
+local DEFAULT_PAGE_SIZE = 512
+
+
+function _M.new(strategy)
+  local self = {
+    strategy = strategy,
+  }
+
+  return setmetatable(self, _MT)
+end
+
+
+local function get_all_nodes_with_sync_cap()
+  local res, err = kong.db.clustering_data_planes:page(DEFAULT_PAGE_SIZE)
+  if err then
+    return nil, "unable to query DB " .. err
+  end
+
+  if not res then
+    return EMPTY
+  end
+
+  local ret = {}
+  local ret_n = 0
+
+  for _, row in ipairs(res) do
+    for _, c in ipairs(row.rpc_capabilities) do
+      if c == "kong.sync.v2" then
+        ret_n = ret_n + 1
+        ret[ret_n] = row.id
+        break
+      end
+    end
+  end
+
+  return ret
+end
+
+
+function _M:notify_all_nodes()
+  local latest_version, err = self.strategy:get_latest_version()
+  if not latest_version then
+    ngx_log(ngx_ERR, "can not get the latest version: ", err)
+    return
+  end
+
+  local msg = { default = { new_version = latest_version, }, }
+
+  for _, node in ipairs(get_all_nodes_with_sync_cap()) do
+    local res, err = kong.rpc:call(node, "kong.sync.v2.notify_new_version", msg)
+    if not res then
+      if not err:find("requested capability does not exist", nil, true) then
+        ngx_log(ngx_ERR, "unable to notify new version: ", err)
+      end
+
+    else
+      ngx_log(ngx_DEBUG, "notified ", node, " ", latest_version)
+    end
+  end
+end
+
+
+function _M:entity_delta_writer(row_id, row, name, options, ws_id)
+  local deltas = {
+    {
+      type = name,
+      id = row_id,
+      ws_id = ws_id,
+      row = row,
+    },
+  }
+
+  local res, err = self.strategy:insert_delta(deltas)
+  if not res then
+    self.strategy:cancel_txn()
+    return nil, err
+  end
+
+  res, err = self.strategy:commit_txn()
+  if not res then
+    self.strategy:cancel_txn()
+    return nil, err
+  end
+
+  self:notify_all_nodes()
+
+  return row
+end
+
+
+-- only control plane has these delta operations
+function _M:register_dao_hooks()
+  local function is_db_export(name)
+    local db_export = kong.db[name].schema.db_export
+    return db_export == nil or db_export == true
+  end
+
+  -- common hook functions (pre/fail/post)
+
+  local function pre_hook_func(entity, name, options)
+    if not is_db_export(name) then
+      return true
+    end
+
+    return self.strategy:begin_txn()
+  end
+
+  local function fail_hook_func(err, entity, name)
+    if not is_db_export(name) then
+      return
+    end
+
+    local res, err = self.strategy:cancel_txn()
+    if not res then
+      ngx_log(ngx_ERR, "unable to cancel cancel_txn: ", tostring(err))
+    end
+  end
+
+  local function post_hook_writer_func(row, name, options, ws_id)
+    if not is_db_export(name) then
+      return row
+    end
+
+    return self:entity_delta_writer(row.id, row, name, options, ws_id)
```

Comment:
This function will return a non-table structure (ngx_null, <userdata1>), but the input entity is row, which is a table, this might introduce a runtime error like:

```
rver: kong_admin, request: "DELETE /routes/3b1b5df0-650c-4d2e-b67d-cd2fd14ae41b HTTP/1.1", host: "127.0.0.1:9001"
2024/10/14 14:18:49 [error] 9172#0: *651 [lua] api_helpers.lua:543: handle_error(): ...l-bin/build/kong-dev/share/lua/5.1/lapis/application.lua:440: ./kong/db/dao/init.lua:39: bad argument #1 to 'pairs' (table expected, got userdata)
stack traceback:
        [C]: in function 'pairs'
        ./kong/db/dao/init.lua:39: in function 'remove_nulls'
        ./kong/db/dao/init.lua:1492: in function 'post_crud_event'                                                                      ./kong/db/dao/init.lua:1357: in function 'delete_entity'
        ./kong/api/endpoints.lua:668: in function 'fn'
        ./kong/api/api_helpers.lua:314: in function <./kong/api/api_helpers.lua:292>
```

In DAO:delete() function, the returned entity from `run_hook("dao:delete:post", ...` will become userdata type, the original one is table.

```

    entity, err_t = run_hook("dao:delete:post", entity, self.schema.name, options, ws_id, cascade_entries)
  if not entity then
    return nil, tostring(err_t), err_t
  end

```

Reply:
fixed in https://github.com/Kong/kong/pull/13157/commits/62ee775f07676db0c9e3c417d445c51b5d38a579